### DISCUSSION TITLE
start lotus worker

### DISCUSSION THREAD

**Discussion by de-authority** [filecoin-project/lotus#7548] [Storage Provider]
# start lotus worker

Compare lotus miner. Lotus worker is simpler. we should pay attention to the environmental variables that are important to the sealing process and the setting of the storage path.

## The role lotus worker play

it's a worker for sealing sectors. 

## Worker path configure

- `$LOTUS_WORKER_PATH `  
Lotus seal workers store all the data in the `$LOTUS_WORKER_PATH `. if you don't specify this environment variable. its default location is `~/.lotusworker`

- `$TMPDIR`  
used when sealing. typically we use nvme disk for sealing. make sure this disk has enough space before starting a sealing job.

## Connect to the storage provider

lotus seal worker communicates with storage provider through the environment variable `$MINER_API_INFO`. You can get `MINER_API_INFO` by running this command in the storage provider:

```shell
lotus-miner auth api-info --perm admin
```

then you can export api info in your lotus seal worker:

```shell
export MINER_API_INFO=<TOKEN>:/ip4/<miner_api_address>/tcp/<port>/http
```

replace `<TOKEN>:/ip4/<miner_api_address>/tcp/<port>/http` with your ownn API info. and typically we can export this environment variable to files like `~/.bashrc` or `/etc/profile`

## Set environment variables

Export these environment variables to your system. Some parameters may be optimized, but you'd better not change them at the beginning. If you know how to optimize, you can modify the values of some parameters

```shell
export BELLMAN_CPU_UTILIZATION=0.875      # optimal value depends on exact hardware
export FIL_PROOFS_MAXIMIZE_CACHING=1
export FIL_PROOFS_USE_GPU_COLUMN_BUILDER=1 # when GPU is available
export FIL_PROOFS_USE_GPU_TREE_BUILDER=1   # when GPU is available
export FIL_PROOFS_PARAMETER_CACHE=/fast/disk/folder # > 100GiB!
export FIL_PROOFS_PARENT_CACHE=/fast/disk/folder2   # > 50GiB!
export FIL_PROOFS_USE_MULTICORE_SDR=1
```

And don't forget to change the` ulimit` of your system. see docs [here](https://github.com/filecoin-project/lotus/discussions/6198)

## Specfy task type (optional)

If you run multiple workers and assign different tasks, it will help to optimize and improve efficiency. Usually, we put a precommit1 precommit2 task in one worker and a provecommit task in another worker.

## Start the program

```shell
nohup lotus-worker run  >> worker.log 2>&1 &
```

When you see `Worker registered successfully, waiting for tasks` in your worker log. congratulations, your worker registered to miner successfully.

## Add more storage
If you find you don't have enough disk space for `lotus-worker`,  or you want to add more permanent storage space. it call added by this command:
- For adding sealing disk space

```shell
lotus-worker storage attach --seal <your-seal-path>
```
- For adding permanent storage disk path

```shell
lotus-worker storage attach --store <your-store-path>
```
and don't forget the `help` flag to check more option

## typical questions

1. moving data between multiple workers  
When you have multiple workers. you can find data transfer between precommit1 to precommit2 workers sometimes. it will cost so much time. A pull request which will improve this situation in progress .you can see detail [here](https://github.com/filecoin-project/lotus/pull/7453)

2.  How to stop lotus-worker?  
When running nohup to terminate (there is no stop command) you have to find the PID of the original process and kill it.

and you can see more official docs about how to run lotus seal worker [**here**](https://docs.filecoin.io/mine/lotus/seal-workers/#frontmatter-title)

**Comment by stuberman** [comment]
When running nohup to terminate (there is no stop command) you have to find the PID of the original process and kill it.

**Reply by de-authority** [reply]
Thanks, Stubeman. I add your tips to the doc. and do we have any other thing that can add?

**Reply by stuberman** [reply]
I think we should also include adding more space/repo to the workers - see [this Slack thread.](https://filecoinproject.slack.com/archives/CEGN061C5/p1636996516482000)

**Reply by de-authority** [reply]
Great. @stuberman 
Thanks for reminding me of this. 
I added the `Add more storage` section.  let me know if you find something that needs more complete
